batches: fix publishing changesets to forks

courier-new opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/43153.

To recap from the issue, we were mistakenly publishing commits to the parent/source repo, rather than the fork repo, which would cause the changeset to fail to publish due to not finding the expected branch on the fork. This PR ensures that when the repo is a fork, we are using `repo.Metadata` rather than `repo.Sources` to set the right clone URL to use when pushing commits.

I also noticed the branch badge on preview changesets was missing the fork indicator it's normally supposed to have, AND the Storybook story for it was broken, due to a sloppy migration. 🤦‍♀️ I've fixed up both of those things, restoring the fork indicator to branches:
 
![Screen Shot 2022-10-19 at 7 33 19 PM](https://user-images.githubusercontent.com/8942601/196847665-ff1daaed-39e4-4372-aed2-6867621706d3.png)


## Test plan

Wrote additional unit tests to validate the proper clone URL from metadata is used when the remote repo is a fork. Fixed up Storybook story coverage of fork indicator on preview changesets. Manually validated by opening changesets in forks against 5 different code hosts:
- github.com
- gitlab.com
- gitlab.sgdev.org
- bitbucket.org
- bitbucket.sgdev.org

![Screen Shot 2022-10-19 at 8 09 45 PM](https://user-images.githubusercontent.com/8942601/196848419-2906eea1-ce7d-411c-9e13-7d50f45de05e.png)

I may still need to update the VCR tests' golden data but I'm still figuring that out. 😬 You can absolutely review this in the meantime, though!

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/kr/fix-forking-clone-url/compare-report.html;tab=live_object) for a full comparison between the commits fee73cc and 4e941be or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

courier-new commented on client/web/src/enterprise/batches/preview/list/storyData.ts:
```diff
@@ -381,39 +381,64 @@ export const visibleChangesetApplyPreviewNodeStories = (
             },
         },
     },
-    'Close changeset': {
```
I didn't actually modify these, I just flipped the order of them so they would match the same order as they're created in the `.story.tsx` file. So you can safely disregard this file. 🙂

courier-new commented on enterprise/cmd/frontend/internal/batches/resolvers/changeset_spec.go:
```diff
@@ -234,5 +234,5 @@ func (r *forkTargetResolver) PushUser() bool {
 }
 
 func (r *forkTargetResolver) Namespace() *string {
-	return r.changesetSpec.GetForkNamespace()
```
`GetForkNamespace` returns `nil` if the fork namespace is just the user's default namespace. This is useful in the other places we use this method, but not for our resolver field, since we use this in the UI to determine how to display the branch badge.

eseliger replied:
can you add that as a code comment as well? I'm sure someone will feel smart later to revert this to the nice helper method we got 😬

courier-new replied:
Haha good point, will do. 😄

eseliger commented on enterprise/internal/batches/sources/sources.go:
```diff
@@ -386,26 +390,92 @@ func setBasicAuth(u *vcs.URL, extSvcType, username, password string) error {
+		cloneURLs = &urls
+	}
+
+	if err != nil {
+		return nil, err
 	}
```
i personally would move the err check to where there err can occur since repo.CloneURLs doesn't seem to fail ever

eseliger commented on enterprise/internal/batches/sources/sources.go:
```diff
@@ -386,26 +390,92 @@ func setBasicAuth(u *vcs.URL, extSvcType, username, password string) error {
+}
+
+func extractGithubCloneURLs(repo *github.Repository) *[]string {
+	cloneURLs := []string{repo.URL}
```
same for gitlab actually. Maybe modifying Sources in the fork could handle that?

eseliger replied:
> For GitHub, it's not available on the metadata, and I thought GitHub repos always support cloning over HTTPS, so if we prefer that anyway why bother with the SSH URL?

I think you can disable it for on-prem. Also that would be inconsistent with how we do it elsewhere.

courier-new replied:
Ah got it. Yeah that makes sense.

courier-new commented on enterprise/internal/batches/sources/sources.go:
```diff
@@ -386,26 +390,92 @@ func setBasicAuth(u *vcs.URL, extSvcType, username, password string) error {
+// getCloneURL returns a remote URL for the repo, or the fork identified by its metadata,
+// preferring HTTPS over SSH.
+func getCloneURL(repo *types.Repo, isFork bool) (*vcs.URL, error) {
+	var cloneURLs *[]string
```
No reason, I think it was relevant in an earlier iteration and I just didn't notice when it stopped being relevant. Will update!

courier-new commented on enterprise/internal/batches/sources/sources.go:
```diff
@@ -386,26 +390,92 @@ func setBasicAuth(u *vcs.URL, extSvcType, username, password string) error {
+	return parsedURLs[0], nil
+}
+
+// extractCloneURLsFromFork returns the remote URLs for the fork identified by the repo metadata.
```
> Can we maybe make this comment a bit even more explicit about that we already modify it in the changeset source?

Yes absolutely.

> Also, why are we modifying metadata but not Sources, so this would've been a non-problem in the first place? 🤔 This is kinda reimplementing the CloneURL logic, isn't it?

You know, I had originally considered that too but I assumed `repo.Sources` needed to be preserved for other reasons. I never actually validated that, though. Let me circle back on that assumption.

courier-new commented on enterprise/internal/batches/reconciler/executor.go:
```diff
@@ -185,10 +183,11 @@ func (e *executor) pushChangesetPatch(ctx context.Context) (err error) {
 	}
 
-	pushConf, err := css.GitserverPushConfig(remoteRepo)
+	pushConf, err := css.GitserverPushConfig(remoteRepo, e.spec.IsFork())
```
Yeah exactly! This gets passed to the private `sources` method `getCloneURL`, which uses it [here](https://github.com/sourcegraph/sourcegraph/pull/43214/files?w=1#diff-bb8e7698f8e9e949afb9bb20f0015f11bb1f570ff02082e9ac39e539d828dfe4R399-R404) to switch between using `repo.Sources` if it's not a fork and `repo.Metadata` if it is.

courier-new commented:
Refactored so that we now update the `repo.Sources` to point to the fork at the time of getting the clone. We still need the metadata to be updated, too, because other changeset operations rely on them. But this at least means less redundancy with `repos.CloneURL()` and no more branching logic when we're pushing commits to forks vs. non-forks. 🙂

I've updated the existing tests around `GetUserFork`/`GetNamespaceFork` to validate `repo.Sources` is being updated correctly, but I still wonder if there ought to be more tests around this. Let me know your thoughts. Although I'm intent on getting this merged ASAP, I'm more than willing to follow up. 🙂

LawnGnome reviewed (approved):
This seems reasonable, although I have to caution that I'm fading on this day of jetlag.

One thing I'm not so sure about is the string munging of the clone URLs: should we be extracting the fork clone URLs from the host-specific metadata, instead of doing a string replacement? This probably has its own complications — we'd have to make sure we use the clone URL for the right protocol if there are multiple options — but feels more like the right thing to do in the longer term.

That said, I know we're on the clock here, so I'm happy to merge this and figure that out for the next release.

courier-new commented:
Thanks for the feedback, Adam! Yeah this is the dilemma -- I originally implemented a solution that only looked at the repo metadata, but not every code host type provides both HTTPS and SSH clone URLs on its metadata (GitHub was the primary culprit), and the metadata clone URLs aren't always in the same format as our `Repo.Sources` URLs, either. So either way there was still some string manipulation impurities that were going to need to happen. 😐

courier-new commented on enterprise/internal/batches/sources/sources.go:
```diff
@@ -433,17 +437,68 @@ func GetRemoteRepo(
+
+	for urn, src := range repo.Sources {
+		if src != nil || src.CloneURL != "" {
+			forkURL := strings.Replace(
+				strings.ToLower(src.CloneURL),
+				strings.ToLower(nameAndOwner),
+				strings.ToLower(forkNameAndOwner),
+				1,
+			)
```
Yeah as far as I could tell all the code hosts treat the clone URL as case-insensitive. I'm happy to revisit this in a follow-up if we think of something better. 🙂
